initial public doc for Distant JS 1.1 (#2)
Kandy-IO/kandy-distant-js: step-by-step/connections.md

# Connections

Now that we have our application and our remote application, it's time to make them communicate so they can cooperate.

## Messages

> The messages used by the 2 parts of the application are application defined. Meaning it's up to the user of Distant to define the format of the message.

Reviewer: ```suggestion
The messages used by the 2 parts of the application are application defined, meaning it's up to the user of Distant to define the format of the message.
```

We suggest using a standard encoding such as JSON and a standard structure for your messages such as [Flux Standard Actions](https://github.com/redux-utilities/flux-standard-action)

> Then we can start defining messages that our remote application will respond to and perform operation.

Reviewer: ```suggestion
Then we can start defining messages that our remote application will respond to by performing operations.
```

Effectively, the `jsonCodec.encode` line is equivalent to the following encode function:

```javascript
function encode(message) {
  const encoder = new TextEncoder()
  return encoder.encode(JSON.stringify(message))
}
```

> Future: In the future we expect to make this easier by allowing simple JavaScript objects to be passed directly by default.

## Receiving from the remote

From the remote application, receiving the message is quite simple.

```javascript
import * as messageTypes from './applicationDefinedMessageTypes'
import remoteController from '@distant/distant-remote'
import { createDistantJSONCodec } from '@distant/distant-codecs'

const jsonCodec = createDistantJSONCodec()

remoteContoller.on('message', encoded => {
  // Message received from the controller.

  // First let's decode it.
  const message = jsonCodec.decode(encoded)

  // Then let's handle it.
  switch (message.type) {
    case messageTypes.MOVE_WINDOW:
      const { bounds } = message.payload
      remoteController.moveWindow(bounds)
      break
    default:
      console.log(`Unknown message type received "${message.type}"`)
      break
  }
})
```

> There we go, we received the message, decoded it, and then performed an operation.

Reviewer: ```suggestion
There we go! We received the message, decoded it, and then performed an operation.
```

> But what if the remote wants to communicate back to the application it's new state?

Reviewer: ```suggestion
But what if the remote wants to communicate its new state back to the application?
```

  const encoded = jsonCodec.encode(message)
  controller.sendMessage(encoded)
}

remoteContoller.on('message', encoded => {
  // Message received from the controller.

  // First let's decode it.
  const message = jsonCodec.decode(encoded)

  // Then let's handle it.
  switch (message.type) {
    case messageTypes.MOVE_WINDOW:
      const { bounds } = message.payload
      remoteController.moveWindow(bounds)

      // Now let's report our new state
      const message = messages.createWindowBoundsStateMessage(remoteController.getWindowBounds)
      sendMessage(remoteController, message)

      break
    default:
      console.log(`Unknown message type received "${message.type}"`)
      break
  }
})
```

Notice the new lines we added? They are highlighted in the diagram. Sending a message from the controller or the session is handled very much in the same way.

We omit some implementation details of `messages.createWindowBoundsStateMessage` but you can imagine that it looks very much like the `createMoveWindowMessage` we saw a few sections above.

Now, we're almost there. Let's complete the picture.

## Receiving messages in the controller application

> In the controller application to receive a message from the session is quite simple and looks very similar to the remote side receiving a message.

Reviewer: ```suggestion
For the controller application to receive a message from the session is quite simple and looks very similar to how the remote side receives a message.
```
